pkg/logging/kinesis/create.go

```diff
@@ -0,0 +1,108 @@
+package kinesis
+
+import (
+	"io"
+
+	"github.com/fastly/cli/pkg/common"
+	"github.com/fastly/cli/pkg/compute/manifest"
+	"github.com/fastly/cli/pkg/config"
+	"github.com/fastly/cli/pkg/errors"
+	"github.com/fastly/cli/pkg/text"
+	"github.com/fastly/go-fastly/v2/fastly"
+)
+
+// CreateCommand calls the Fastly API to create Amazon Kinesis logging endpoints.
+type CreateCommand struct {
+	common.Base
+	manifest manifest.Data
+
+	// required
+	EndpointName string // Can't shaddow common.Base method Name().
+	Version      int
+	StreamName   string
+	AccessKey    string
+	SecretKey    string
+	Region       string
+
+	// optional
+	Format            common.OptionalString
+	FormatVersion     common.OptionalUint
+	ResponseCondition common.OptionalString
+	Placement         common.OptionalString
+}
+
+// NewCreateCommand returns a usable command registered under the parent.
+func NewCreateCommand(parent common.Registerer, globals *config.Data) *CreateCommand {
+	var c CreateCommand
+	c.Globals = globals
+	c.manifest.File.Read(manifest.Filename)
+	c.CmdClause = parent.Command("create", "Create an Amazon Kinesis logging endpoint on a Fastly service version").Alias("add")
+
+	c.CmdClause.Flag("name", "The name of the Kinesis logging object. Used as a primary key for API access").Short('n').Required().StringVar(&c.EndpointName)
+	c.CmdClause.Flag("service-id", "Service ID").Short('s').StringVar(&c.manifest.Flag.ServiceID)
+	c.CmdClause.Flag("version", "Number of service version").Required().IntVar(&c.Version)
+
+	c.CmdClause.Flag("stream-name", "The Amazon Kinesis stream to send logs to").Required().StringVar(&c.StreamName)
+	c.CmdClause.Flag("access-key", "The access key associated with the target Amazon Kinesis stream").Required().StringVar(&c.AccessKey)
+	c.CmdClause.Flag("secret-key", "The secret key associated with the target Amazon Kinesis stream").Required().StringVar(&c.SecretKey)
+
+	c.CmdClause.Flag("region", "The AWS region where the Kinesis stream exists").Required().StringVar(&c.Region)
```

Comment:
Yeah, it is required. The configuration validation will fail if it is not present. I see what you mean in the case of `go-fastly`. I followed the existing patterns of other logging endpoints when I did that work and they all seem to only have `ServiceId` and `ServiceVersion` as required. My guess is that the enforcement of required fields was deferred to the config validation, but maybe we want to revisit that.

Reply:
**Question**:
> The configuration validation will fail if it is not present.

I presume 'config validation' refers to the upstream API's own validation, would that be correct?

Reply:
I see Go-Fastly currently doesn't define validation for the 'region' field, but if we're saying the API itself is expecting a region field otherwise it'll return an error, then we should ensure a validation check for the 'region' field is added into Go-Fastly (I have an existing PR open I can add this to).

But for the time being, this is fine to be marked as `.Required()` considering Go-Fastly doesn't _currently_ have logic to catch the missing field.

> Note: I can't actually test whether the API expects 'region' to be provided because it seems Kinesis logging isn't available on my personal Fastly account 😬 I did login as an admin to see if it's something I could enable as a feature flag but nothing obvious was coming up in my search.

Reply:
Also, just FYI (most of which will be obvious to you): my perspective on this is that we have three tools at play here:

```
CLI > Go-Fastly > API
```

We should be able to trust the API to always have the right validation in place, but (as you know) by having validation _outside_ of the API it means we can prevent slowing down a client trying to carry out operations that otherwise we know are just going to fail by the time they reach the API layer, so short-circuiting those requests is better for the client as far as the 'feedback loop' is concerned but also good for the API because it doesn't have to handle the request in the first place.

Considering Go-Fastly and CLI sit at effectively the same level (i.e. they exist _on_ the client) we need to decide which 'tool' is responsible for handling validation of obvious issues with a client request, and for me that should be Go-Fastly as it's the thing that's ultimately fronting the API.

So my hope is that with the work @doramatadora  has been doing on the new [OpenAPI specification](https://github.com/fastly/api-documentation/pull/433) it will mean we can move to a place where we more concretely know what is a required field and have Go-Fastly handle validation and thus reduce duplicating logic in the CLI (although admittedly at this point in time the OpenAPI spec [for Kinesis logging](https://github.com/fastly/api-documentation/pull/433/files#diff-56bc7bba0f7cc7c57afed506447d0e9d30eff59cc148624f3177437f225e0d3a) still doesn't indicate that 'region' is a required field nor does it show it as part of its example code so that would likely need updating).

Reply:
> I presume 'config validation' refers to the upstream API's own validation, would that be correct?

That's correct.

> Note: I can't actually test whether the API expects 'region' to be provided because it seems Kinesis logging isn't available on my personal Fastly account grimacing I did login as an admin to see if it's something I could enable as a feature flag but nothing obvious was coming up in my search.

If you send me the service id in slack I can enable this for you.

I'll also follow-up on the docs side of things and open a change request if needed to have it indicate `region` as required.